src/python/espressomd/particle_data.py

```diff
@@ -515,10 +515,9 @@ def bonds(self):
 
         """
         bonds = []
-        for bond_view in self.call_method("get_bonds_view"):
-            bond_id = bond_view[0]
-            partner_ids = bond_view[1:]
-            bonds.append((BondedInteractions()[bond_id], *partner_ids))
+        for bond_id, *partner_ids in self.call_method("get_bonds_view"):
+            bond = self.call_method("get_bond_by_id", bond_id=bond_id)
+            bonds.append((bond, *partner_ids))
```

Comment:
We want to completely hide bond ids from the Python interface. The main driving force was recurring bugs in the `ScriptInterface`. Python bond ids could diverge from core bond ids during checkpointing, or refer to a different object when users erased and created a bond with the same type and same id as the erased one. Users could also run into subtle Python-specific oddities when making assumptions about data structures, e.g. tuples vs lists in https://github.com/espressomd/espresso/issues/4391#issuecomment-967064220.

Hiding bond ids also allows us to change the internal data structure from a `std::unordered_map` to a container with a more favorable time complexity for element search, which could be relevant in polymer melt simulations where there are a lot of bonds. Right now we are using a hash function, which scales as $\mathcal{O}(1)$ on average, but $\mathcal{O}(N)$ in worst case. We could imagine using a `std::vector`, which requires a multiplication instead of a hash, or a hybrid container that would use an array for small storage and `std::unordered_map` for large storage.

Lastly, we are very close to having multi-system capability in ESPResSo. A bond id is not useful if you don't know in which system this id is registered.

Reply:
What I meant was only that it looked like the return type of `self.call_method("get_bonds_view")` is `bond_id, partner_id_0, partner_id_1,...`, so the size of this changes if there are several partners involved. To me this looked confusing, why not all the partner_ids are returned in a vector

Reply:
Probably a relic from the Tcl days. This is part of the API, so we cannot really change it at the moment. At some point we will store bonds on all particles, rather than the first one; that project will give us an opportunity to explore more intuitive APIs.